File: 108.convert-sorted-array-to-binary-search-tree.py
#
# @lc app=leetcode id=108 lang=python3
#
# [108] Convert Sorted Array to Binary Search Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        if (len(nums)==0):
            return None
        length = len(nums)
        middleIndex = length//2
        leftSubtree = nums[:middleIndex]
        rightSubtree = nums[middleIndex+1:]
        leftRoot = self.sortedArrayToBST(leftSubtree)
        rightRoot = self.sortedArrayToBST(rightSubtree)
        
        root = TreeNode(val=nums[middleIndex], left=leftRoot, right=rightRoot)
        return root
        
# @lc code=end

File: 108.convert-sorted-array-to-binary-search-tree.go
/*
 * @lc app=leetcode id=108 lang=golang
 *
 * [108] Convert Sorted Array to Binary Search Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
package main

/*
recursive
base case: if len is zero
general case: get the middle one and make it a node, finish left tree and right tree and concat them, return root
*/
func sortedArrayToBST(nums []int) *TreeNode {
	if len(nums) == 0 {
		return nil
	}
	l := 0
	r := len(nums) - 1
	m := (l + r) / 2
	root := TreeNode{
		Val:   nums[m],
		Left:  sortedArrayToBST(nums[:m]),
		Right: sortedArrayToBST(nums[m+1:]),
	}
	return &root
}

// @lc code=end
